pkg/sfu/bwe/remotebwe/remote_bwe.go

```diff
@@ -0,0 +1,391 @@
+	var estimateToCommit int64
+	switch reason {
+	case channelCongestionReasonLoss:
+		estimateToCommit = int64(float64(r.lastExpectedBandwidthUsage) * (1.0 - r.params.Config.NackRatioAttenuator*r.channelObserver.GetNackRatio()))
+	default:
+		estimateToCommit = r.lastReceivedEstimate
+	}
+	if estimateToCommit > r.lastReceivedEstimate {
+		estimateToCommit = r.lastReceivedEstimate
+	}
+
+	commitThreshold := int64(r.params.Config.ExpectedUsageThreshold * float64(r.lastExpectedBandwidthUsage))
+	action := "applying"
+	if estimateToCommit > commitThreshold {
+		action = "skipping"
+	}
+
+	if action == "applying" {
+		r.params.Logger.Infow(
+			fmt.Sprintf("remote bwe: channel congestion detected, %s channel capacity update", action),
+			"reason", reason,
+			"old(bps)", r.committedChannelCapacity,
+			"new(bps)", estimateToCommit,
+			"lastReceived(bps)", r.lastReceivedEstimate,
+			"expectedUsage(bps)", r.lastExpectedBandwidthUsage,
+			"commitThreshold(bps)", commitThreshold,
+			"channel", r.channelObserver.ToString(),
+		)
+	} else {
+		r.params.Logger.Debugw(
+			fmt.Sprintf("remote bwe: channel congestion detected, %s channel capacity update", action),
+			"reason", reason,
+			"old(bps)", r.committedChannelCapacity,
+			"new(bps)", estimateToCommit,
+			"lastReceived(bps)", r.lastReceivedEstimate,
+			"expectedUsage(bps)", r.lastExpectedBandwidthUsage,
+			"commitThreshold(bps)", commitThreshold,
+			"channel", r.channelObserver.ToString(),
+		)
+	}
+	/*
+		r.params.Logger.Debugw(
+			fmt.Sprintf("remote bwe: channel congestion detected, %s channel capacity: experimental", action),
+			"nackHistory", r.channelObserver.GetNackHistory(),
+		)
+	*/
+	if estimateToCommit > commitThreshold {
+		return false
+	}
+
+	r.committedChannelCapacity = estimateToCommit
+
+	// reset to get new set of samples for next trend
+	r.channelObserver = r.newChannelObserverNonProbe()
+	return true
+}
+
+func (r *RemoteBWE) updateCongestionState(state bwe.CongestionState, reason channelCongestionReason) {
+	r.params.Logger.Infow(
+		"remote bwe: congestion state change",
+		"from", r.congestionState,
+		"to", state,
+		"reason", reason,
+		"committedChannelCapacity", r.committedChannelCapacity,
+	)
+
+	if state != r.congestionState {
+		// notify worker for ticker interval management based on state
+		select {
+		case r.wake <- struct{}{}:
+		default:
+		}
+	}
+
+	r.congestionState = state
+	r.congestionStateSwitchedAt = mono.Now()
+}
+
+func (r *RemoteBWE) newChannelObserverNonProbe() *channelObserver {
+	return newChannelObserver(
+		channelObserverParams{
+			Name:   "non-probe",
+			Config: r.params.Config.ChannelObserverNonProbe,
+		},
+		r.params.Logger,
+	)
+}
+
+func (r *RemoteBWE) ProbingStart(expectedBandwidthUsage int64) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.isInProbe = true
+	r.lastExpectedBandwidthUsage = expectedBandwidthUsage
+
+	channelState := ""
+	if r.channelObserver != nil {
+		channelState = r.channelObserver.ToString()
```

Comment:
`stringer` interface?

Reply:
Will change these. Sorry, I picked up that habit from somewhere when I started using Go and that has just stuck. Don't remember where I saw it and picked up.